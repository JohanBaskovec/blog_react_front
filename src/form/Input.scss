@import "./FormComponent.scss";
.Input {
  display: flex;
  width: 100%;

  @extend .FormComponent;
}
