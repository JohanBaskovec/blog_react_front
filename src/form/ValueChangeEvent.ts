export class ValueChangeEvent<T> {
    constructor(private _name: string, private _newValue: T) {

    }

    get name(): string {
        return this._name;
    }

    get newValue(): T {
        return this._newValue;
    }
}

export type ValueChangeEventHandler<T> = (e: ValueChangeEvent<T>) => void;
