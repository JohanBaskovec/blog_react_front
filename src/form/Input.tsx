import React from "react";
import "./Input.scss";
import {ValueChangeEvent, ValueChangeEventHandler} from "./ValueChangeEvent";

export interface InputProps {
    onValueChange?: ValueChangeEventHandler<string>;
    onChange?: React.ChangeEventHandler<HTMLInputElement>;
    style?: React.CSSProperties;
    name: string;
    value: string;
}

export function Input(props: InputProps): JSX.Element {
    const onChange = (event: React.ChangeEvent<HTMLInputElement>) => {
        if (props.onValueChange) {
            props.onValueChange(new ValueChangeEvent<string>(props.name, event.target.value));
        }
        if (props.onChange) {
            props.onChange(event);
        }
    };
    return (<input name={props.name}
                   value={props.value}
                   onChange={onChange}
                   id={props.name}
                   style={props.style}
                   className="Input"/>);
}
