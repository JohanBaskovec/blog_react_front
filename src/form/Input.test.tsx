import React from "react";
import TestRenderer from 'react-test-renderer';
import {fireEvent, render, screen} from '@testing-library/react'
import '@testing-library/jest-dom/extend-expect'
import {Input} from "./Input";
import {ValueChangeEvent} from "./ValueChangeEvent";

describe('Input', () => {
    it("works", () => {
        let onChangeWorks = false;
        const onChange = (e: React.ChangeEvent) => {
            onChangeWorks = true;
        }
        let value = '';
        let onValueChangeEvent: ValueChangeEvent<string> | null = null;
        const onValueChange = (e: ValueChangeEvent<string>) => {
            value = e.newValue;
            onValueChangeEvent = e;
        }
        const name = 'inputName';
        const element = <Input name={name} value={value} onChange={onChange} onValueChange={onValueChange}/>;
        render(element);
        const input: HTMLElement = screen.getByRole('textbox');
        const text = 'test';
        fireEvent.input(input, {target: {value: text}});

        expect(onChangeWorks).toBe(true);
        expect(value).toBe(text);
        expect(onValueChangeEvent).not.toBeNull();
        expect(onValueChangeEvent!.name).toBe(name);
    });
    it("is rendered properly", () => {
        const tree = TestRenderer
            .create(<Input value='read-only' name='name'/>)
            .toJSON();
        expect(tree).toMatchSnapshot();
    });
});
