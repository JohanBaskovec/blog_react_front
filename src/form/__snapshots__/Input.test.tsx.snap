// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input is rendered properly 1`] = `
<input
  className="Input"
  id="name"
  name="name"
  onChange={[Function]}
  value="read-only"
/>
`;
